crates/services/gas_price_service/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+use async_trait::async_trait;
+use fuel_core_services::{
+    RunnableService,
+    RunnableTask,
+    ServiceRunner,
+    StateWatcher,
+};
+use fuel_core_types::fuel_types::BlockHeight;
+use std::sync::Arc;
+
+use tokio::sync::{
+    RwLock,
+    RwLockReadGuard,
+};
+
+pub mod static_updater;
+
+pub fn new_service<A, U>(
+    current_fuel_block_height: BlockHeight,
+    update_algo: U,
+) -> anyhow::Result<ServiceRunner<GasPriceService<A, U>>>
+where
+    U: UpdateAlgorithm<Algorithm = A> + Send + Sync,
+    A: Send + Sync,
+{
+    let service = GasPriceService::new(current_fuel_block_height, update_algo);
+    Ok(ServiceRunner::new(service))
+}
+
+pub struct GasPriceService<A, U> {
+    next_block_algorithm: BlockGasPriceAlgo<A>,
+    update_algorithm: U,
+}
+
+impl<A, U> GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A>,
+    A: Send + Sync,
+{
+    pub fn new(starting_block_height: BlockHeight, update_algorithm: U) -> Self {
+        let algorithm = update_algorithm.start(starting_block_height);
+        let next_block_algorithm = BlockGasPriceAlgo::new(algorithm);
+        Self {
+            next_block_algorithm,
+            update_algorithm,
+        }
+    }
+
+    pub fn next_block_algorithm(&self) -> BlockGasPriceAlgo<A> {
+        self.next_block_algorithm.clone()
+    }
+}
+
+#[async_trait]
+pub trait UpdateAlgorithm {
+    type Algorithm;
+
+    fn start(&self, for_block: BlockHeight) -> Self::Algorithm;
+    async fn next(&mut self) -> Self::Algorithm;
+}
+
+impl<A, U> GasPriceService<A, U>
+where
+    U: UpdateAlgorithm<Algorithm = A>,
+    A: Send + Sync,
+{
+    async fn update(&mut self, new_algorithm: A) {
+        self.next_block_algorithm.update(new_algorithm).await;
+    }
+}
+
+#[derive(Debug)]
+pub struct BlockGasPriceAlgo<A>(Arc<RwLock<A>>);
```

Comment:
This wrapper type seems unnecessary and a bit misleading. It's just `Arc<RwLock<_>>`, and contains no gas price -related functionality by itself.

Reply:
The name `Block` is outdated. I'll change that.

Maybe `SharedGasPriceAlgo`.

Ideally, we'd have two types:
`GasPriceAlgoWriter` and `GasPriceAlgoReader` and then protect them so that mutability is only controlled by the service, but our services are over-constrained with the `SharedData` assoc type, so it kinda makes us use the same type here and on the provider. At least `update` is `&mut self` 🤷

Reply:
I think I understand the tradeoff here. I don't think the layer of mental indirection is worth it. Since this disagreement seems to be a just matter of taste, I'll just approve it.